vitest can't be placed in the first level catalog

### Describe the bug

If workdir is `/root/app`, then it will work fine. [(example)](https://github.com/ViZhe/vitest-bug-docker/runs/4821304385?check_suite_focus=true)
But with `/root` or `/app` vitest will throw the error [(example)](https://github.com/ViZhe/vitest-bug-docker/runs/4821304336?check_suite_focus=true)

Error:
```
node:internal/process/promises:246
          triggerUncaughtException(err, true /* fromPromise */);
          ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/node_modules/local-pkg/index.mjs' imported from /root/node_modules/vitest/dist/worker.js
    at new NodeError (node:internal/errors:371:5)
    at finalizeResolution (node:internal/modules/esm/resolve:416:11)
    at moduleResolve (node:internal/modules/esm/resolve:932:10)
    at defaultResolve (node:internal/modules/esm/resolve:1044:11)
    at ESMLoader.resolve (node:internal/modules/esm/loader:422:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:222:40)
    at ESMLoader.import (node:internal/modules/esm/loader:276:22)
    at importModuleDynamically (node:internal/modules/esm/translators:111:35)
    at importModuleDynamicallyCallback (node:internal/process/esm_loader:35:14)
    at interpretedImport (file:///root/node_modules/vitest/dist/worker.js:130:15)
The command '/bin/sh -c npm i install && npm run test:run' returned a non-zero code: 1
```

### Reproduction

Repo:
https://github.com/ViZhe/vitest-bug-docker
Action with error:
https://github.com/ViZhe/vitest-bug-docker/runs/4821304336?check_suite_focus=true

### System Info

```shell
System:
    OS: Linux 5.10 Alpine Linux
    CPU: (4) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 154.16 MB / 1.94 GB
    Container: Yes
    Shell: 1.34.1 - /bin/ash
  Binaries:
    Node: 16.13.1 - /usr/local/bin/node
    Yarn: 1.22.15 - /usr/local/bin/yarn
    npm: 8.1.2 - /usr/local/bin/npm
  npmPackages:
    vite: latest => 2.7.12 
    vitest: latest => 0.1.13
```


### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite-node/src/utils.ts
+++ b/packages/vite-node/src/utils.ts
@@ -27,7 +27,7 @@ export function isPrimitive(v: any) {
 export function toFilePath(id: string, root: string): string {
   let absolute = slash(id).startsWith('/@fs/')
     ? id.slice(4)
-    : id.startsWith(dirname(root))
+    : id.startsWith(dirname(root)) && dirname(root) !== '/'
       ? id
       : id.startsWith('/')
         ? slash(resolve(root, id.slice(1)))

--- a/test/core/test/file-path.test.ts
+++ b/test/core/test/file-path.test.ts
@@ -1,87 +1,80 @@
-import { afterEach, describe, expect, it } from 'vitest'
+import { describe, expect, it, vi } from 'vitest'
 import { isWindows, slash, toFilePath } from '../../../packages/vite-node/src/utils'
 
-const originalCwd = process.cwd()
-
-function mockCwd(cwd: string) {
-  Object.defineProperty(process, 'cwd', {
-    value: () => cwd,
-  })
-}
-
-function restoreCwd() {
-  Object.defineProperty(process, 'cwd', {
-    value: originalCwd,
-  })
-}
-
 describe('toFilePath', () => {
-  afterEach(() => {
-    restoreCwd()
-  })
-
+  // the following tests will work incorrectly on unix systems
   if (isWindows) {
-    // these test will work incorrectly on unix systems
     it('windows', () => {
       const root = 'C:/path/to/project'
-      mockCwd(root)
-
       const id = '/node_modules/pkg/file.js'
       const expected = 'C:/path/to/project/node_modules/pkg/file.js'
+
+      const processSpy = vi.spyOn(process, 'cwd').mockReturnValue(root)
       const filePath = toFilePath(id, root)
+      processSpy.mockRestore()
+
       expect(slash(filePath)).toEqual(expected)
     })
 
     it('windows with /@fs/', () => {
       const root = 'C:/path/to/project'
-      mockCwd(root)
-
       const id = '/@fs/C:/path/to/project/node_modules/pkg/file.js'
       const expected = 'C:/path/to/project/node_modules/pkg/file.js'
+
+      const processSpy = vi.spyOn(process, 'cwd').mockReturnValue(root)
       const filePath = toFilePath(id, root)
+      processSpy.mockRestore()
+
       expect(slash(filePath)).toEqual(expected)
     })
   }
 
+  // the following tests will work incorrectly on windows systems
   if (!isWindows) {
-    // these test will work incorrectly on windows systems
     it('unix', () => {
       const root = '/path/to/project'
-      mockCwd(root)
-
       const id = '/node_modules/pkg/file.js'
       const expected = '/path/to/project/node_modules/pkg/file.js'
+
+      const processSpy = vi.spyOn(process, 'cwd').mockReturnValue(root)
       const filePath = toFilePath(id, root)
+      processSpy.mockRestore()
+
       expect(slash(filePath)).toEqual(expected)
     })
 
     it('unix with /@fs/', () => {
       const root = '/path/to/project'
-      mockCwd(root)
-
       const id = '/@fs//path/to/project/node_modules/pkg/file.js'
       const expected = '/path/to/project/node_modules/pkg/file.js'
+
+      const processSpy = vi.spyOn(process, 'cwd').mockReturnValue(root)
       const filePath = toFilePath(id, root)
+      processSpy.mockRestore()
       expect(slash(filePath)).toEqual(expected)
     })
 
     it('unix in first level catalog', () => {
       const root = '/root'
-      mockCwd(root)
-
       const id = '/node_modules/pkg/file.js'
       const expected = '/root/node_modules/pkg/file.js'
+
+      const processSpy = vi.spyOn(process, 'cwd').mockReturnValue(root)
       const filePath = toFilePath(id, root)
+      processSpy.mockRestore()
+
       expect(slash(filePath)).toEqual(expected)
     })
 
     it('unix with /@fs/ in first level catalog', () => {
       const root = '/root'
-      mockCwd(root)
-
       const id = '/@fs//root/node_modules/pkg/file.js'
       const expected = '/root/node_modules/pkg/file.js'
+
+      const processSpy = vi.spyOn(process, 'cwd').mockReturnValue(root)
       const filePath = toFilePath(id, root)
+      processSpy.mockRestore()
+
       expect(slash(filePath)).toEqual(expected)
     })
   }

--- a/test/core/test/file-path.test.ts
+++ b/test/core/test/file-path.test.ts
@@ -1,0 +1,88 @@
+import { afterEach, describe, expect, it } from 'vitest'
+import { isWindows, slash, toFilePath } from '../../../packages/vite-node/src/utils'
+
+const originalCwd = process.cwd()
+
+function mockCwd(cwd: string) {
+  Object.defineProperty(process, 'cwd', {
+    value: () => cwd,
+  })
+}
+
+function restoreCwd() {
+  Object.defineProperty(process, 'cwd', {
+    value: originalCwd,
+  })
+}
+
+describe('toFilePath', () => {
+  afterEach(() => {
+    restoreCwd()
+  })
+
+  if (isWindows) {
+    // these test will work incorrectly on unix systems
+    it('windows', () => {
+      const root = 'C:/path/to/project'
+      mockCwd(root)
+
+      const id = '/node_modules/pkg/file.js'
+      const expected = 'C:/path/to/project/node_modules/pkg/file.js'
+      const filePath = toFilePath(id, root)
+      expect(slash(filePath)).toEqual(expected)
+    })
+
+    it('windows with /@fs/', () => {
+      const root = 'C:/path/to/project'
+      mockCwd(root)
+
+      const id = '/@fs/C:/path/to/project/node_modules/pkg/file.js'
+      const expected = 'C:/path/to/project/node_modules/pkg/file.js'
+      const filePath = toFilePath(id, root)
+      expect(slash(filePath)).toEqual(expected)
+    })
+  }
+
+  if (!isWindows) {
+    // these test will work incorrectly on windows systems
+    it('unix', () => {
+      const root = '/path/to/project'
+      mockCwd(root)
+
+      const id = '/node_modules/pkg/file.js'
+      const expected = '/path/to/project/node_modules/pkg/file.js'
+      const filePath = toFilePath(id, root)
+      expect(slash(filePath)).toEqual(expected)
+    })
+
+    it('unix with /@fs/', () => {
+      const root = '/path/to/project'
+      mockCwd(root)
+
+      const id = '/@fs//path/to/project/node_modules/pkg/file.js'
+      const expected = '/path/to/project/node_modules/pkg/file.js'
+      const filePath = toFilePath(id, root)
+      expect(slash(filePath)).toEqual(expected)
+    })
+
+    it('unix in first level catalog', () => {
+      const root = '/root'
+      mockCwd(root)
+
+      const id = '/node_modules/pkg/file.js'
+      const expected = '/root/node_modules/pkg/file.js'
+      const filePath = toFilePath(id, root)
+      expect(slash(filePath)).toEqual(expected)
+    })
+
+    it('unix with /@fs/ in first level catalog', () => {
+      const root = '/root'
+      mockCwd(root)
+
+      const id = '/@fs//root/node_modules/pkg/file.js'
+      const expected = '/root/node_modules/pkg/file.js'
+      const filePath = toFilePath(id, root)
+      expect(slash(filePath)).toEqual(expected)
+    })
+  }
+})

